### PR TITLE
Create HuggingFaceTransformer.py

### PR DESCRIPTION
**Pull request by mobashgr**
Here is the code for adding any HuggingFace Transformer model over INCEpTION

### CODE REVIEW

File: ariadne/contrib/HuggingFaceTransformer.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        model = AutoModelForTokenClassification.from_pretrained(self._model)
+        nlp_ner = pipeline("ner", model=model, tokenizer=tokenizer,aggregation_strategy="max")
+        for c, sentence in enumerate(cas.select(SENTENCE_TYPE)):
+            columns = {'word', 'start', 'end', 'entity_group', 'score'}
+            df = pd.DataFrame(columns=columns)
+            s=nlp_ner(sentence.get_covered_text())
+            for item in s:
+                 df = df.append(item, ignore_index= True)
+            for i in range(len(df)):
+               prediction = create_prediction(cas, layer, feature,df.loc[i, "start"]+sentence.__getattribute__("begin"),df.loc[i, "end"]+sentence.__getattribute__("begin"),df.loc[i, "entity_group"])
```
Comment (reckart) [inline ariadne/contrib/HuggingFaceTransformer.py:28]: Do you really need the `__getattribute__` here? Actually, `sentence.begin` should work.

File: ariadne/contrib/HuggingFaceTransformer.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    '''As an example, to use it HuggingFace models for token classification  
+    HuggingFaceClassifier(model_name="ghadeermobasher/BC5CDR-Chemical-Disease-balanced-pubmedbert")'''
+    def __init__(self, model_name: str):
+            super().__init__()
+            self._model = model_name
+    def predict(self, cas: Cas, layer: str, feature: str, project_id: str, document_id: str, user_id: str):
+        tokenizer = AutoTokenizer.from_pretrained(self._model)
+        model = AutoModelForTokenClassification.from_pretrained(self._model)
+        nlp_ner = pipeline("ner", model=model, tokenizer=tokenizer,aggregation_strategy="max")
+        for c, sentence in enumerate(cas.select(SENTENCE_TYPE)):
```
Comment (reckart) [inline ariadne/contrib/HuggingFaceTransformer.py:21]: I can't see the `c` being used. If it is not needed, I guess the `enumerate` is not needed either?

Comment (mobashgr) [reply ariadne/contrib/HuggingFaceTransformer.py:21]: Yes, true. I was using them for other purposes and forgot to remove them.

**Comment by jcklie** [comment]
Thank you for the PR! The basics looks good to me.
This code still has the issue that you do not use the tokenization of INCEpTION, which requires you then to use character level granularity which is not so nice to use later when exporting the corpus and using it downstream. In other recommenders, we align the predictions of recommenders to the INCEpTION tokenization which you would need to do here also before I would merge it tbh. Examples and hints can be found in 

https://github.com/huggingface/transformers/issues/14305
https://huggingface.co/docs/transformers/custom_datasets?highlight=offset_mapping#token-classification-with-wnut-emerging-entities
https://discuss.huggingface.co/t/predicting-with-token-classifier-on-data-with-no-gold-labels/9373

I also do not understand why you would need pandas here, it is certainly possible to do it just without.
As you only support token classification here, I would also name it TransformerTokenClassifier or so, the name indicates that it is a general implementation.

**Comment by mobashgr** [comment]
Regarding the first point, yes, I was facing this problem yesterday and used the character level granularity as suggested by Richard. My problem was resolved, and I don’t think that I have the time to do this alignment now. I just wanted to share what I have as a solution to a problem that I was facing especially since the Adapter code isn’t working, and it was misleading TBH.  I believe that INCEpTION is a very powerful tool and it should definitely have examples for HuggingFace classifiers. 

For the second point, I need pandas, as the output of the pipeline in my case is a list of list of dictionaries. A sample of the pipeline output looks like this `[{'entity_group': 'Chemical', 'score': 0.9996301, 'word': 'acety', 'start': 66, 'end': 71}, {'entity_group': 'Chemical', 'score': 0.99999845, 'word': 'nicotine', 'start': 98, 'end': 106}, {'entity_group': 'Chemical', 'score': 0.99911577, 'word': 'la dicine evised', 'start': 122, 'end': 144}, {'entity_group': 'Chemical', 'score': 0.9999038, 'word': 'alpha - only hete', 'start': 308, 'end': 325}]` . So, I prefer to change it into a dataframe.

**Comment by reckart** [comment]
@mobashgr Sorry for getting back to you late. Could you please add the same Apache License license header to the file that we use in the other files?

I believe it should not be a strong problem if the recommender users a different tokenization. If the recommender creates a suggestion that does not fit in with the layer settings in INCEpTION, it will be ignored - it should not cause trouble.

**Comment by reckart** [comment]
I believe this is superseded by https://github.com/inception-project/inception-external-recommender/pull/63

**Pull request closed without merging**